### ISSUE TITLE
`Callable` seems not a priority in generic inference

### ISSUE THREAD

**Issue by panhongtao07** [microsoft/pyright#4848]
**Describe the bug**
When a generic type is unioned with `Callable`, the `Callable` is not considered in the inference.

**To Reproduce**
In the following code:
```python
from typing import Callable, TypeVar, reveal_type, overload

T = TypeVar('T')

def get_type_hint(func: Callable[[], T] | Callable[[T], None] | list[T] | dict[str, T] | T) -> T:
    return func # type: ignore

def get_type_hint_no_other(func: Callable[[], T]) -> T:
    return func # type: ignore

def from_str(val: str) -> None: ...
def to_str() -> str:
    return ''

reveal_type(get_type_hint(['']))            # str
reveal_type(get_type_hint({'': 1}))         # int
reveal_type(get_type_hint(from_str))        # error: (val: str) -> None
reveal_type(get_type_hint(to_str))          # error: () -> str
reveal_type(get_type_hint(str))             # error: Type[str]
reveal_type(get_type_hint_no_other(str))    # str
```
`mypy` has the same problem in the `Callable` inference.

**Expected behavior**

The `reveal_type` for calling `get_type_hint` with `from_str`, `to_str` and `str` should be `str`.

I know this is an [Ambiguous Solution Scoring](https://microsoft.github.io/pyright/#/mypy-comparison?id=constraint-solver-ambiguous-solution-scoring) problem and both answers are correct. But I think the `Callable` should be considered before the unlimited generic type.

Since the `Callable` is a first-class type in Python, it should be considered with same priority as other types, such as `list`, `dict` and `tuple`. However, other types are considered in the inference, but `Callable` is not.

**VS Code extension or command-line**
Pyright 1.1.300

**Additional context**
May I ask how to visit the typing community? I am not sure if this is the right place to report this issue and other ideas.

**Comment by panhongtao07** [comment]
In this case, we can use `overload` to solve the problem:

```python
from typing import Callable, TypeVar, reveal_type

T = TypeVar('T')

@overload
def get_type_hint(func: Callable[[], T]) -> T: ...
@overload
def get_type_hint(func: T) -> T: ...

reveal_type(get_type_hint(to_str))  # str
reveal_type(get_type_hint(str))     # str
```

But this is not a good solution. We have to double the number of overloads.

**Comment by panhongtao07** [comment]
Thanks. I'm wondering if there is a way to avoid the ambiguity. I do think `Callable | T` is not elegant, but I can't find a better way to express it. It represents a function that accepts a value or a value factory.

If TypeVar can add exclusive constraints, it would be better. For example, we can add a constraint that the type is not `Callable`:

```python
T = TypeVar('T')
T_not_callable = TypeVar('T_not_callable', exclude=Callable)
T_not_callable = TypeVar('T_not_callable', excludes=[Callable])
T_not_callable: TypeAlias = T - Callable  # or T / Callable
```

Using `-` because type variables and callables represent a set of types, and `-` is the set difference operator, which is the same as `exclude` and is prior to `|` in the operator precedence. Just a syntax sugar.

Then we can use `T_not_callable` in the union:

```python
def get_type_hint(func: Callable[[], T] | T_not_callable) -> T:
    return func # type: ignore

def get_type_hint(func: Callable[[], T] | T - Callable) -> T:
    return func # type: ignore
```

But I believe it's hard to implement, since we have to add extra fields to `TypeVar` and maybe other classes, modify the `TypeVar` constuctor, and benefit few occasions.

**Comment by erictraut** [comment]
What you're proposing is an extension to the Python type system, not something that an individual type checker like pyright would implement independently of the broader Python ecosystem. Your proposal also would not compose with existing features of the type system, so it's unlikely that it would ever be accepted by the Python typing community.

I'll look at improving the heuristics, which are still in bounds with the existing type system.

**Comment by panhongtao07** [comment]
Of course. So I said the last sentence. I didn’t expect this less common feature to be implemented. I was just confused about what an appropriate solution would be.

**Comment by panhongtao07** [comment]
How, that's great! Thanks for the quick fixes!

**Comment by erictraut** [comment]
This is included in pyright 1.1.301, which I just published. It will also be included in a future release of pylance.
